### PR TITLE
Fixed Extra dot in multi filename #3641

### PR DESCRIPTION
**Pull request by gbargsley**
## Type of Change

 - [X] Bug fix (non-breaking change, fixes #<enter issue number>)
 - [ ] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 

### Purpose
 
Fix Issue 3641.  Not sure if best method, but it works.
### Screenshots

image

### CODE REVIEW

**Comment by niphlod** [comment]
wan't the problem an EXTRA dot ?

**Comment by niphlod** [comment]
(BTW, if it fixes a bug, put the number in the description...... @wsmelton spent a lot of time in templating GH issues)

**Comment by niphlod** [comment]
How can adding a dot in code solve the problem of having an extra dot in
the filename? Ghggggh, I'm prolly missing something

Il mer 30 mag 2018, 15:25 Garry Bargsley <notifications@github.com> ha
scritto:

**Comment by gbargsley** [comment]
I removed a dot in the source code that was between the filename and extention.  I think somehow the $suffix was getting an a dot in it, and made it have two dots.  So I just removed the hardcoded dot as I could not tell where $suffix was getting a dot added.

**Comment by niphlod** [comment]
okay, misread in mobile the diff, sorry.

**Comment by niphlod** [comment]
something else seems to be the culprit, let's see if I can replicate

**Comment by niphlod** [comment]
line 336 is the culprit...... a fileinfo extension property contains the actual dot.

**Comment by gbargsley** [comment]
I saw that, but did not think the .extension would add a dot.  So should we strip it from that call to make for better code? Then add the dot back to where it was in the line 362?

**Comment by gbargsley** [comment]
I did this to line 336: `$suffix = $file.extension.Replace(".","")`
And added the period back to 362: `$FinalBackupPath[$i] = $FinalBackupPath[$i] + "\" + $($File.BaseName) + "-$($i+1)-of-$FileCount.$suffix"`

And it seems to work.

**Comment by niphlod** [comment]
yup, fix would just be changing that line to 
`$suffix = $file.Extension  -Replace '^\.', ''`

BTW: tests failed correctly with your solution, maybe slap in a regr test to make sure #3461 doesn't happen again

**Comment by gbargsley** [comment]
I changed the code to your way.  Unfortunately, I do not know what you mean about a regression test.  I would love to learn.

**Comment by niphlod** [comment]
I recently added, e.g., https://github.com/sqlcollaborative/dbatools/blob/development/tests/Backup-DbaDatabase.Tests.ps1#L126 to make sure all backups are done with extension (till recently there were cases where extension was not used). A regression test is a test that makes sure a problem won't pop up ever again. In this case, the side-case of running a backup with:
- a single backuppath
- a single backupfilename 
- filecount greater than 1 
showed that files with more than one dot were created.
The regression test in this case would be to create lines in the pester test that make sure that when you run backup-dbadatabase with:
- a single backuppath
- a single backupfilename 
- filecount greater than 1 
created files have no the double-dot in them.

**Comment by gbargsley** [comment]
Ok, let me try.

**Comment by gbargsley** [comment]
So all green.  Does that mean the test I added worked?

**Comment by niphlod** [comment]
seems good . BTW: isn't 
`[IO.Path]::GetExtension($path) | Should -Not -BeLike '*..*'` more readable ?

**Comment by gbargsley** [comment]
But what if 3 or anything other than 1 is used in the filename/extension?

**Comment by niphlod** [comment]
3 or 1 of what ? tests are not "dynamic". 
You ask for a backup with a filename with .bak extension and you expect the resulting files to not have more than one dot (i.e. "it doesn't change the extension passed in"). 
It'd be cool to have genetic test cases (i.e. computers checking random inputs and verifying all things obey to rules) but, alas, we're not there yet.

**Comment by gbargsley** [comment]
I meant if it gets ... or .... or more. Your solution seems to look for only ..

**Comment by niphlod** [comment]
My solution looks for `*..*`. which we deemed wrong. 
Coincidentally, `*..*` does not only match two consecutive periods but also also `...` (3 consecutive periods) and `....` (4 consecutive periods)  and so on till the end of path limitations in PS so ..... I don't see where your doubts arise

**Comment by gbargsley** [comment]
It is not doubt as I would never attempt to do that. I am just trying to understand.

**Comment by niphlod** [comment]
Match is "two or more" so all cases are covered.

Il mer 30 mag 2018, 19:00 Garry Bargsley <notifications@github.com> ha
scritto:

**Comment by potatoqualitee** [comment]
thank you both!

**Comment by wsmelton** [comment]
@gbargsley when you do your commits if you include specific term/phrase as:

fixes #3641 

It will cause the issue to be closed once Chrissy merges it. I should also do the same thing if you put that phrase in the PR content too.

**Comment by gbargsley** [comment]
So not in the title and only in the description if I use the #. 

Trying to understand.

**Comment by niphlod** [comment]
https://help.github.com/articles/closing-issues-using-keywords/ . 
tl;dr: fill the template sqlcollaborative has already
